Cargo los datos sobre temperaturas durante el año 2024 de Barcelona desde la estación de Drassanes que tiene un criterio de temperatura a nivel ciudad, aunque en el caso de los datos existentes en este año esta estación no contiene los relacionados con la velocidad media del viento, por lo que busco la estación más cercana a Drassanes que si los contenga y encuentro la del Observatorio Fabra.

In [20]:
import pandas as pd

# Cargo los archivos CSV
df1 = pd.read_csv('temp_012024_062024.csv')
df2 = pd.read_csv('temp_072024_122024.csv')
df3 = pd.read_csv('fabra1.csv')
df4 = pd.read_csv('fabra2.csv')

# Concateno los DataFrames por filas
df_temp_2024 = pd.concat([df1, df2], axis=0, ignore_index=True)
df_B = pd.concat([df3, df4], axis=0, ignore_index=True)

df_temp_2024.head()

fecha indicativo                nombre  provincia  altitud  tmed prec  \
0  2024-01-01      0201X  BARCELONA, DRASSANES  BARCELONA       11  11,8  0,0   
1  2024-01-02      0201X  BARCELONA, DRASSANES  BARCELONA       11  11,6  0,0   
2  2024-01-03      0201X  BARCELONA, DRASSANES  BARCELONA       11  15,4  0,0   
3  2024-01-04      0201X  BARCELONA, DRASSANES  BARCELONA       11  14,2  0,4   
4  2024-01-05      0201X  BARCELONA, DRASSANES  BARCELONA       11  11,6  8,8   

   tmin horatmin  tmax horatmax  hrMedia  hrMax horaHrMax  hrMin horaHrMin  
0   8,8    23:52  14,7    14:45     68.0   76.0     05:00   61.0     10:50  
1   7,9    03:53  15,3    15:59     46.0   66.0     00:00   27.0     16:00  
2  12,6    07:23  18,2    12:22     61.0   71.0     07:40   50.0     01:30  
3  11,6    07:02  16,9    13:18     79.0   85.0     08:40   65.0     00:00  
4   9,6   Varias  13,6    12:46     79.0   89.0     04:30   67.0     13:30

El dataframe A está compuesto por 365 filas y 20 columnas

In [21]:
df_temp_2024.shape

(366, 16)

Hay columnas que no contemplo utilizar por lo que decido eliminarlas para aligerar el dataframe.

In [22]:
df_temp_2024= df_temp_2024.drop(columns=['indicativo', 'nombre', 'horatmin', 'horatmax', 'altitud', 'hrMax', 'hrMin', 'horaHrMin', 'horaHrMax'])

df_temp_2024.head()

fecha  provincia  tmed prec  tmin  tmax  hrMedia
0  2024-01-01  BARCELONA  11,8  0,0   8,8  14,7     68.0
1  2024-01-02  BARCELONA  11,6  0,0   7,9  15,3     46.0
2  2024-01-03  BARCELONA  15,4  0,0  12,6  18,2     61.0
3  2024-01-04  BARCELONA  14,2  0,4  11,6  16,9     79.0
4  2024-01-05  BARCELONA  11,6  8,8   9,6  13,6     79.0

El dataframe con los datos de viento del Observatorio Fabra contiene 366 filas y 21 columnas.

In [23]:
df_B.shape

(366, 21)

Observo el detalle de las columnas existentes.

In [24]:
df_B.head()

fecha indicativo            nombre  provincia  altitud  tmed prec  \
0  2024-01-01      0200E  BARCELONA, FABRA  BARCELONA      408  11,0  0,0   
1  2024-01-02      0200E  BARCELONA, FABRA  BARCELONA      408  10,4  0,0   
2  2024-01-03      0200E  BARCELONA, FABRA  BARCELONA      408  13,6   Ip   
3  2024-01-04      0200E  BARCELONA, FABRA  BARCELONA      408  13,4  0,9   
4  2024-01-05      0200E  BARCELONA, FABRA  BARCELONA      408   8,8  8,7   

   tmin horatmin  tmax  ... dir  velmedia racha horaracha  sol presMax  \
0   7,6    06:34  14,5  ...  27       1,9   9,4     23:40  4,9   971,3   
1   6,9   Varias  13,9  ...  27       4,4  12,2     03:19  3,1   971,0   
2  11,5    23:59  15,6  ...  26       3,3  13,9     09:16  0,2   968,1   
3   9,7    07:46  17,2  ...  27       1,7   7,8     22:52  3,5   969,0   
4   5,7    23:59  11,9  ...  29       3,3  10,6     23:31  0,1   959,0   

  horaPresMax presMin horaPresMin hrMedia  
0          22   968,2          05    67.0  
1          00   965,9          24    41.0  
2          23   964,3          04    65.0  
3          10   959,0          24    80.0  
4          00   953,6          14    83.0  

[5 rows x 21 columns]

Elimino la mayoría de las columnas, manteniendo únicamente la velocidad del viento, lal cual agregaré al primer dataframe.

In [25]:
df_B= df_B.drop(columns=['fecha', 'provincia', 'indicativo', 'nombre', 'altitud', 'tmed', 'prec', 'tmin','horatmin', 'tmax', 'horatmax', 'dir', 'racha', 'horaracha', 'sol', 'presMax', 'horaPresMax', 'presMin', 'horaPresMin', 'hrMedia'])

df_B.head()

velmedia
0      1,9
1      4,4
2      3,3
3      1,7
4      3,3

In [26]:
df_temp_2024['velmedia']= df_B['velmedia']
df_temp_2024.head()

fecha  provincia  tmed prec  tmin  tmax  hrMedia velmedia
0  2024-01-01  BARCELONA  11,8  0,0   8,8  14,7     68.0      1,9
1  2024-01-02  BARCELONA  11,6  0,0   7,9  15,3     46.0      4,4
2  2024-01-03  BARCELONA  15,4  0,0  12,6  18,2     61.0      3,3
3  2024-01-04  BARCELONA  14,2  0,4  11,6  16,9     79.0      1,7
4  2024-01-05  BARCELONA  11,6  8,8   9,6  13,6     79.0      3,3

Reviso las características de cada columna, observo varios detalles sobre ellas que considero tratar:

* `fecha` es del tipo `object` y el formato debería ser `datetime`
* Hay varias columas que aparecen como tipo `object` al utiizar una coma como decimal y las transformaré como float con punto.


In [27]:
df_temp_2024.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   fecha      366 non-null    object 
 1   provincia  366 non-null    object 
 2   tmed       365 non-null    object 
 3   prec       366 non-null    object 
 4   tmin       365 non-null    object 
 5   tmax       365 non-null    object 
 6   hrMedia    361 non-null    float64
 7   velmedia   366 non-null    object 
dtypes: float64(1), object(7)
memory usage: 23.0+ KB


Realizo la transformación de la columna `fecha` al tipo `datetime`

In [28]:
df_temp_2024['fecha']= pd.to_datetime(df_temp_2024['fecha'], format='%Y-%m-%d')
print(df_temp_2024.dtypes)

fecha        datetime64[ns]
provincia            object
tmed                 object
prec                 object
tmin                 object
tmax                 object
hrMedia             float64
velmedia             object
dtype: object


Transformo las columnas 'tmed', 'prec', 'tmin', 'tmax' al tipo float, modificando el símbolo de coma por punto.

In [29]:
columnas_a_float = ['tmed', 'prec', 'tmin', 'tmax', 'velmedia']

# Reemplazo las comas por puntos para que pueda realizarse el cambio al tipo float
for col in columnas_a_float:
    df_temp_2024[col] = df_temp_2024[col].str.replace(',', '.')  

df_temp_2024[columnas_a_float] = df_temp_2024[columnas_a_float].astype(float)

print(df_temp_2024.dtypes)

fecha        datetime64[ns]
provincia            object
tmed                float64
prec                float64
tmin                float64
tmax                float64
hrMedia             float64
velmedia            float64
dtype: object


Compruebo la existencia de valores nulos

In [30]:
df_temp_2024.isnull().sum()

fecha        0
provincia    0
tmed         1
prec         0
tmin         1
tmax         1
hrMedia      5
velmedia     0
dtype: int64

Analizo los valores nulos de la variable `hrMedia`; observo que aparecen varias columnas en ese registro con valores faltantes.

In [31]:
nulos_tmed = df_temp_2024['hrMedia'].isnull()
df_temp_2024[nulos_tmed]

fecha  provincia  tmed  prec  tmin  tmax  hrMedia  velmedia
17  2024-01-18  BARCELONA  18.0   0.4  15.0  21.1      NaN       6.7
34  2024-02-04  BARCELONA  14.8   0.0   9.5  20.2      NaN       1.1
219 2024-08-07  BARCELONA  28.1   0.0  24.5  31.7      NaN       2.8
225 2024-08-13  BARCELONA  28.3   1.0  25.5  31.1      NaN       1.4
364 2024-12-30  BARCELONA  10.4   0.0   5.5  15.4      NaN       1.4

La idea de la imputación en este caso es mantener la continuidad de la serie. Imputo los valores faltantes utilizando una función, que en caso de que encuentre un valor nulo en una columna, buscará el valor correspondiente en la fila anterior (día anterior). Si el valor anterior existe, se utiliza para reemplazar el valor nulo. 

In [32]:
# Ordeno el DataFrame por fecha para asegurar la correcta imputación
df_temp_2024 = df_temp_2024.sort_values('fecha')

# Selecciono las columnas a imputar 
columnas_a_imputar = ['tmed', 'tmin', 'tmax', 'hrMedia']

# Se itera sobre las filas y las columnas a imputar
for index, row in df_temp_2024.iterrows():
    for columna in columnas_a_imputar:
        if pd.isna(row[columna]):
            # Buscar el valor del día anterior en la misma columna
            valor_anterior = df_temp_2024.loc[index - 1, columna]
            # Imputar si el valor anterior existe
            if not pd.isna(valor_anterior):
                df_temp_2024.loc[index, columna] = valor_anterior

Compruebo que los valores ha sido imputados totalmente.

In [33]:
df_temp_2024.isnull().sum()

fecha        0
provincia    0
tmed         0
prec         0
tmin         0
tmax         0
hrMedia      0
velmedia     0
dtype: int64

- Renombro las columna `provincia` por `ciudad`.

In [34]:
df_temp_2024= df_temp_2024.rename(columns={'provincia': 'ciudad', 'hrMedia': 'hrmedia'})
df_temp_2024= df_temp_2024[['fecha', 'ciudad', 'tmin', 'tmax', 'tmed', 'prec','hrmedia', 'velmedia']]
df_temp_2024.head()


fecha     ciudad  tmin  tmax  tmed  prec  hrmedia  velmedia
0 2024-01-01  BARCELONA   8.8  14.7  11.8   0.0     68.0       1.9
1 2024-01-02  BARCELONA   7.9  15.3  11.6   0.0     46.0       4.4
2 2024-01-03  BARCELONA  12.6  18.2  15.4   0.0     61.0       3.3
3 2024-01-04  BARCELONA  11.6  16.9  14.2   0.4     79.0       1.7
4 2024-01-05  BARCELONA   9.6  13.6  11.6   8.8     79.0       3.3

Codifico la columna `ciudad` utilizando el valor `1` para poder relacionarlo en la base de datos con la tabla que corresponda a las ciudades

In [35]:
mapeo_ciudad= {"BARCELONA": "1"}

df_temp_2024['ciudad']= df_temp_2024['ciudad'].map(mapeo_ciudad)
df_temp_2024.head()

fecha ciudad  tmin  tmax  tmed  prec  hrmedia  velmedia
0 2024-01-01      1   8.8  14.7  11.8   0.0     68.0       1.9
1 2024-01-02      1   7.9  15.3  11.6   0.0     46.0       4.4
2 2024-01-03      1  12.6  18.2  15.4   0.0     61.0       3.3
3 2024-01-04      1  11.6  16.9  14.2   0.4     79.0       1.7
4 2024-01-05      1   9.6  13.6  11.6   8.8     79.0       3.3

Exporto el dataframe a un archivo en formato `csv`

In [36]:
df_temp_2024.to_csv('temp_bcn_2024.csv', index= False, encoding='utf-8')

Resumen del dataframe creado:

- Se eliminaron las columnas 'indicativo', 'nombre', 'hrMax', 'horaHrMax', 'hrMin', 'horaHrMin', 'altitud', 'dir', 'racha', 'horaracha', 'horatmin', 'horatmax'.
- Se obtuvo el valor de velocidad media desde dos csv de otra estación de la ciudad que se agregó al dataframe 
- Se realizaron transformaciones de tipo en las columnas 'tmed', 'prec', 'tmin', 'tmax', 'velmedia'.
- Se transformó la columna 'fecha' a tipo datetime
- Se trataron los valores nulos mediante la imputación del registro del día anterior al faltante, buscando mantener la continuidad de la serie. No hay valores faltantes.
- Se renombró la columna 'provincia' por 'ciudad' y se asignó el valor 1 para que sea utilizado en un modelo de tablas relaciones como identificador(id).
- Se exporto el dataframe a un archivo en formato csv.